### PR TITLE
LUCENE-8086

### PR DESCRIPTION
**Pull request by iverase**
Here are the changes, in particular:

- Geo3dFactory: Use GeoExactCircle for non-spherical planets.
- Geo3dCircleShape: Remove method relate.
- Geo3DShape: Use new factory method for building GeoBbox from bounds object.
- Geo3dDistanceCalculator: use pointonbearing from planet model.
- Test refactoring

### CODE REVIEW

File: lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeRectRelationTestCase.java [hunk 1]
```diff
@@ -16,29 +16,15 @@
...
-import org.apache.lucene.spatial3d.geom.GeoPolygon;
+import org.junit.Rule;
+import org.junit.Test;
 import org.locationtech.spatial4j.TestLog;
 import org.locationtech.spatial4j.context.SpatialContext;
-import org.locationtech.spatial4j.distance.DistanceUtils;
 import org.locationtech.spatial4j.shape.Circle;
 import org.locationtech.spatial4j.shape.Point;
 import org.locationtech.spatial4j.shape.RectIntersectionTestHelper;
-import org.apache.lucene.spatial3d.geom.LatLonBounds;
```
Comment (dsmiley) [inline lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeRectRelationTestCase.java]: If I get this right, you've removed the Geo3D dependencies of this test.  Yet it's still named to be Geo3d?

File: lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dCircleShape.java [hunk 2]
```diff
@@ -67,16 +64,4 @@ public Point getCenter() {
     }
     return center;
   }
-
```
Comment (dsmiley) [inline lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dCircleShape.java]: Yay

File: lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dDistanceCalculator.java [hunk 3]
```diff
@@ -73,62 +74,20 @@ public boolean within(Point from, double toX, double toY, double distance) {
...
-    // Algorithm using Vincenty's formulae (https://en.wikipedia.org/wiki/Vincenty%27s_formulae)
-    // which takes into account that planets may not be spherical.
-    //Code adaptation from http://www.movable-type.co.uk/scripts/latlong-vincenty.html
     Geo3dPointShape geoFrom = (Geo3dPointShape) from;
     GeoPoint point = (GeoPoint) geoFrom.shape;
-    double lat = point.getLatitude();
-    double lon = point.getLongitude();
     double dist = DistanceUtils.DEGREES_TO_RADIANS * distDEG;
     double bearing = DistanceUtils.DEGREES_TO_RADIANS * bearingDEG;
-
```
Comment (dsmiley) [inline lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dDistanceCalculator.java]: Yay

File: lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dShapeFactory.java [hunk 4]
```diff
@@ -67,6 +74,25 @@ public SpatialContext getSpatialContext() {
...
+   * surface circle and planes that describe the circle. Therefore on WSG84, since the
+   * radius of earth is 6,371,000 meters, an accuracy of 1e-6 corresponds to 6.3 meters.
+   * For an accuracy of 1.0 meters, the value of 1.6e-7.
+   *
+   * The default value is set to 10m (1.6e-6).
+   *
+   * Note that accuracy has no effect when the planet model is a sphere. In that case circles
+   * are always fully precise.
+   *
+   * @param circleAccuracy the provided accuracy as a linear distance.
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dShapeFactory.java]: I need to ask Karl Wright if that is what it means, but I guess so. I will update accordingly.

File: lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dShapeFactory.java [hunk 5]
```diff
@@ -150,10 +176,21 @@ public Rectangle rect(double minX, double maxX, double minY, double maxY) {
 
   @Override
   public Circle circle(double x, double y, double distance) {
-    GeoCircle circle = GeoCircleFactory.makeGeoCircle(planetModel,
-        y * DistanceUtils.DEGREES_TO_RADIANS,
-        x * DistanceUtils.DEGREES_TO_RADIANS,
-        distance * DistanceUtils.DEGREES_TO_RADIANS);
+    GeoCircle circle;
+    if (planetModel.ab == planetModel.c) {
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dShapeFactory.java]: I think spatial3d is a low level library in that respect, so it shouldn't have such a method. Karl Wright has the last word, comment would be fine

File: lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dShapeFactory.java [hunk 6]
```diff
@@ -55,6 +55,13 @@
   private SpatialContext context;
   private PlanetModel planetModel;
 
+  /**
+   * Default accuracy for circles when not using the unit sphere.
+   * It is equivalent to 10m on the surface of the earth.
+   */
+  private static double DEFAULT_CIRCLE_ACCURACY = 1.6e-6;
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dShapeFactory.java]: Indeed!

File: lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeRectRelationTestCase.java [hunk 7]
```diff
@@ -155,16 +107,12 @@ protected Geo3dShape generateRandomShape(Point nearP) {
           ulhcPoint = lrhcPoint;
           lrhcPoint = temp;
         }
-        final GeoBBox shape = GeoBBoxFactory.makeGeoBBox(planetModel, ulhcPoint.getY() * DEGREES_TO_RADIANS,
-            lrhcPoint.getY() * DEGREES_TO_RADIANS,
-            ulhcPoint.getX() * DEGREES_TO_RADIANS,
-            lrhcPoint.getX() * DEGREES_TO_RADIANS);
-        return new Geo3dShape(shape, ctx);
+        return (Geo3dShape<?>) ctx.getShapeFactory().rect(lrhcPoint, ulhcPoint);
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeRectRelationTestCase.java]: Indeed!

File: lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeRectRelationTestCase.java (hunk 1 above)
Comment (iverase) [reply lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeRectRelationTestCase.java]: I didn't dear to change it but that was the idea of the effort, I will remove Geo3d.

File: lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/ShapeRectRelationTestCase.java [hunk 8]
```diff
@@ -75,9 +75,9 @@ public void testGeoCircleRect() {
     new Geo3dRectIntersectionTestHelper(ctx) {
```
Comment (dsmiley) [inline lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/ShapeRectRelationTestCase.java]: Should be renamed as this class is no longer Geo3d specific

**Pull request closed without merging**